File: cli/cmd/project.go
package cmd

import (
	"fmt"
	"os"
	"text/tabwriter"

	"connectrpc.com/connect"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	v1 "github.com/kendricklawton/liquid-metal/gen/go/liquidmetal/v1"
	v1connect "github.com/kendricklawton/liquid-metal/gen/go/liquidmetal/v1/liquidmetalv1connect"
)

var projectCmd = &cobra.Command{
	Use:   "project",
	Short: "Manage projects",
}

var projectListCmd = &cobra.Command{
	Use:   "list",
	Short: "List projects in the active workspace",
	RunE:  runProjectList,
}

var projectUseCmd = &cobra.Command{
	Use:   "use <slug-or-id>",
	Short: "Set project_id in ./liquid-metal.toml",
	Args:  cobra.ExactArgs(1),
	RunE:  runProjectUse,
}

func init() {
	projectCmd.AddCommand(projectListCmd, projectUseCmd)
}

func runProjectList(_ *cobra.Command, _ []string) error {
	t := requireToken()
	workspaceID := viper.GetString("workspace_id")
	if workspaceID == "" {
		return fmt.Errorf("no active workspace — run `flux workspace use <slug>` first")
	}

	// Show the project_id currently in liquid-metal.toml (if any) as the active marker.
	activeCfg := viper.New()
	activeCfg.SetConfigName("liquid-metal")
	activeCfg.SetConfigType("toml")
	activeCfg.AddConfigPath(".")
	_ = activeCfg.ReadInConfig()
	activeProject := activeCfg.GetString("service.project_id")

	client := v1connect.NewProjectServiceClient(newHTTPClient(), apiURL(), connect.WithGRPC())
	resp, err := client.ListProjects(cmdCtx(), withToken(connect.NewRequest(&v1.ListProjectsRequest{
		WorkspaceId: workspaceID,
	}), t))
	if err != nil {
		return fmt.Errorf("list projects: %w", err)
	}

	projects := resp.Msg.GetProjects()
	if len(projects) == 0 {
		fmt.Println("No projects found in this workspace.")
		return nil
	}

	w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	fmt.Fprintln(w, "  SLUG\tNAME\tID")
	for _, p := range projects {
		marker := "  "
		if p.GetId() == activeProject {
			marker = "* "
		}
		fmt.Fprintf(w, "%s%s\t%s\t%s\n",
			marker,
			p.GetSlug(),
			p.GetName(),
			p.GetId(),
		)
	}
	w.Flush()
	return nil
}

func runProjectUse(_ *cobra.Command, args []string) error {
	t := requireToken()
	target := args[0]
	workspaceID := viper.GetString("workspace_id")
	if workspaceID == "" {
		return fmt.Errorf("no active workspace — run `flux workspace use <slug>` first")
	}

	client := v1connect.NewProjectServiceClient(newHTTPClient(), apiURL(), connect.WithGRPC())
	resp, err := client.ListProjects(cmdCtx(), withToken(connect.NewRequest(&v1.ListProjectsRequest{
		WorkspaceId: workspaceID,
	}), t))
	if err != nil {
		return fmt.Errorf("list projects: %w", err)
	}

	var projectID string
	for _, p := range resp.Msg.GetProjects() {
		if p.GetId() == target || p.GetSlug() == target {
			projectID = p.GetId()
			break
		}
	}
	if projectID == "" {
		return fmt.Errorf("project %q not found in active workspace", target)
	}

	// Update (or create) ./liquid-metal.toml with the new project_id.
	if err := setProjectInToml(projectID); err != nil {
		return fmt.Errorf("update liquid-metal.toml: %w", err)
	}

	fmt.Printf("Set project_id in liquid-metal.toml: %s\n", projectID)
	return nil
}

// setProjectInToml writes project_id into ./liquid-metal.toml, preserving other keys.
func setProjectInToml(projectID string) error {
	cfg := viper.New()
	cfg.SetConfigName("liquid-metal")
	cfg.SetConfigType("toml")
	cfg.AddConfigPath(".")
	_ = cfg.ReadInConfig() // OK if file doesn't exist yet

	cfg.Set("service.project_id", projectID)

	// If no config file was loaded, write a new one.
	if cfg.ConfigFileUsed() == "" {
		cfg.SetConfigFile("liquid-metal.toml")
	}
	return cfg.WriteConfig()
}

File: crates/cli/src/commands/project.rs
use anyhow::Result;
use serde::Deserialize;

use crate::client::ApiClient;
use crate::config::Config;

#[derive(Deserialize)]
struct Project {
    id: String,
    name: String,
    slug: String,
}

pub async fn run_list(config: &Config) -> Result<()> {
    let token = config.require_token()?;
    let workspace_id = config
        .workspace_id
        .as_deref()
        .ok_or_else(|| anyhow::anyhow!("no active workspace — run `flux workspace use <slug>` first"))?;

    let client = ApiClient::new(config.api_url(), Some(token));
    let projects: Vec<Project> = client
        .get(&format!("/projects?workspace_id={}", workspace_id))
        .await?;

    if projects.is_empty() {
        println!("No projects found in this workspace.");
        return Ok(());
    }

    let active_project = read_project_toml_id();
    println!("  {:<20} {:<20} {}", "SLUG", "NAME", "ID");
    for p in &projects {
        let marker = if active_project.as_deref() == Some(&p.id) { "* " } else { "  " };
        println!("{}{:<20} {:<20} {}", marker, p.slug, p.name, p.id);
    }
    Ok(())
}

pub async fn run_use(config: &Config, slug_or_id: &str) -> Result<()> {
    let token = config.require_token()?;
    let workspace_id = config
        .workspace_id
        .as_deref()
        .ok_or_else(|| anyhow::anyhow!("no active workspace — run `flux workspace use <slug>` first"))?;

    let client = ApiClient::new(config.api_url(), Some(token));
    let projects: Vec<Project> = client
        .get(&format!("/projects?workspace_id={}", workspace_id))
        .await?;

    let project = projects
        .iter()
        .find(|p| p.id == slug_or_id || p.slug == slug_or_id)
        .ok_or_else(|| anyhow::anyhow!("project {:?} not found in active workspace", slug_or_id))?;

    set_project_in_toml(&project.id)?;
    println!("Set project_id in liquid-metal.toml: {}", project.id);
    Ok(())
}

fn read_project_toml_id() -> Option<String> {
    let contents = std::fs::read_to_string("liquid-metal.toml").ok()?;
    let val: toml::Value = toml::from_str(&contents).ok()?;
    val.get("service")?.get("project_id")?.as_str().map(|s| s.to_string())
}

fn set_project_in_toml(project_id: &str) -> Result<()> {
    let mut val: toml::Value = if let Ok(contents) = std::fs::read_to_string("liquid-metal.toml") {
        toml::from_str(&contents)?
    } else {
        toml::Value::Table(toml::map::Map::new())
    };

    val.as_table_mut()
        .ok_or_else(|| anyhow::anyhow!("invalid liquid-metal.toml"))?
        .entry("service")
        .or_insert(toml::Value::Table(toml::map::Map::new()))
        .as_table_mut()
        .ok_or_else(|| anyhow::anyhow!("invalid [service] section"))?
        .insert("project_id".to_string(), toml::Value::String(project_id.to_string()));

    std::fs::write("liquid-metal.toml", toml::to_string(&val)?)?;
    Ok(())
}
